### PR TITLE
FMCDAQ3 Ultrascale Support

### PR DESCRIPTION
**Pull request by adrimbarean**
- Add Ultrascale and Ultrascale+ support for FMCDAQ3 and by extension to Microblaze projects.
 - Fix SPI related bug in the FMCDAQ3 project.
 - Do minor refactoring of the code.

### CODE REVIEW

File: common_drivers/xcvr_core/xcvr_core.c [hunk 1]
```diff
@@ -81,82 +81,107 @@ int32_t xcvr_write(xcvr_core *core,
 #ifdef XILINX
 
 /***************************************************************************//**
- * @brief xcvr_drp_read Dynamic reconfiguration port read access for Xilinx devices
```
Comment (commodo) [inline common_drivers/xcvr_core/xcvr_core.c]: this commit seems to do too much at once

File: common_drivers/xcvr_core/xcvr_core.c [hunk 2]
```diff
@@ -228,73 +228,93 @@ int32_t xcvr_setup(xcvr_core *core)
...
-
-			xilinx_xcvr_qpll_write_config(core, &qpll_config);
+			ret |= xilinx_xcvr_calc_qpll_config(core, core->ref_rate_khz,
+					core->lane_rate_khz, &qpll_config, &core->dev.out_div);
+			out_div = core->dev.out_div;
+
+			ln_rate_kbps = xilinx_xcvr_qpll_calc_lane_rate(core,
+					core->ref_rate_khz,
+					&qpll_config,
+					out_div);
```
Comment (stefpopa) [inline common_drivers/xcvr_core/xcvr_core.c]: nitpick: you probably can fit this on fewer lines

File: common_drivers/xcvr_core/xcvr_core.c [hunk 3]
```diff
@@ -228,73 +228,93 @@ int32_t xcvr_setup(xcvr_core *core)
...
-
-			xilinx_xcvr_cpll_write_config(core, &cpll_config);
+			ret |= xilinx_xcvr_calc_cpll_config(core, core->ref_rate_khz,
+					core->lane_rate_khz, &cpll_config, &core->dev.out_div);
+			out_div = core->dev.out_div;
+
+			ln_rate_kbps = xilinx_xcvr_cpll_calc_lane_rate(core,
+					core->ref_rate_khz,
+					&cpll_config,
+					out_div);
```
Comment (stefpopa) [inline common_drivers/xcvr_core/xcvr_core.c]: nitpick: you probably can fit this on fewer lines

File: common_drivers/xcvr_core/xcvr_core.c [hunk 4]
```diff
@@ -228,73 +228,93 @@ int32_t xcvr_setup(xcvr_core *core)
...
-			xilinx_xcvr_configure_cdr(core, core->lane_rate_kbps,
-				out_div, core->dev.lpm_enable);
+			xilinx_xcvr_configure_lpm_dfe_mode(core,
+					XCVR_DRP_PORT_COMMON,
+					core->dev.lpm_enable);
+			xilinx_xcvr_configure_cdr(core,
+					XCVR_DRP_PORT_COMMON,
+					core->lane_rate_khz,
+					out_div,
+					core->dev.lpm_enable);
```
Comment (stefpopa) [inline common_drivers/xcvr_core/xcvr_core.c]: nitpick: you probably can fit this on fewer lines

File: common_drivers/xcvr_core/xcvr_core.h [hunk 5]
```diff
@@ -179,9 +179,17 @@ int32_t xcvr_write(xcvr_core *core, uint32_t reg_addr, uint32_t reg_data);
...
+	int32_t xilinx_xcvr_drp_read(xcvr_core *core, uint32_t drp_port,uint32_t reg);
+	int32_t xilinx_xcvr_drp_write(xcvr_core *core,
+			uint32_t drp_port,
+			uint32_t reg,
+			uint32_t val);
+	int32_t xcvr_drp_update(xcvr_core *core,
+			uint32_t drp_port,
+			uint32_t reg,
+			uint32_t mask,
+			uint32_t val);
```
Comment (stefpopa) [inline common_drivers/xcvr_core/xcvr_core.h]: nitpick: you probably can fit this on fewer lines

File: common_drivers/xcvr_core/xcvr_modules/xilinx_qpll.c [hunk 6]
```diff
@@ -49,144 +49,222 @@
...
+	val = xilinx_xcvr_drp_read(core, drp_port, QPLL_FBDIV(qpll));
+	if (val < 0)
+		return val;
+
+	conf->fb_div = (val & 0xff) + 2;
+
+	conf->band = 0;
+
+//	dev_err(core->dev, "qpll: fb_div=%d, qpll: refclk_div=%d\n",
+//			conf->fb_div, conf->refclk_div);
```
Comment (stefpopa) [inline common_drivers/xcvr_core/xcvr_modules/xilinx_qpll.c]: please remove commented code

File: common_drivers/xcvr_core/xcvr_core.c [hunk 7]
```diff
@@ -228,73 +228,93 @@ int32_t xcvr_setup(xcvr_core *core)
 	uint32_t rx_out_div;
 	uint32_t tx_out_div;
 	uint32_t ln_rate_kbps;
-	xcvr_cpll cpll_config;
-	xcvr_qpll qpll_config;
+	struct xilinx_xcvr_cpll_config cpll_config;
```
Comment (stefpopa) [inline common_drivers/xcvr_core/xcvr_core.c]: I still feel that this commit is doing too much. Even if this is not the case, you need to convince the reviewers by describing what you are actually doing in the commit message. What does adding support for Ultrascale and Ultrascale+ imply? What are the differences from previously supported architectures?

**Comment by adrimbarean** [comment]
V2 -> v3:
 - editing out extra lines
 - remove commented code
 - apply Astyle
 - separate and explain commits better

**Comment by adrimbarean** [comment]
V3 -> V4:
 - remove redundant message in QPLL and CPLL commits
 - add description to fmcdaq3 project update commit

**Pull request merged**